Catch failures in add_node

In case there is a failure to add new node (e.g after decommission or terminate_and_replace), we should catch it and:
1. Remove it from cluster.db_nodes list.
2. Retry adding node.
(We can set max number of retries to 3).

```
2019-01-19 05:33:56,228 nemesis          L0448 ERROR| sdcm.nemesis.ChaosMonkey: Exception in random_disrupt_method terminate_and_replace_node: Setup has failed!
2019-01-19 05:33:56,229 nemesis          L0734 ERROR| sdcm.nemesis.ChaosMonkey: Unhandled exception in method <function disrupt at 0x7fc248b01050>
Traceback (most recent call last):
  File "/sct/sdcm/nemesis.py", line 884, in disrupt
  File "/sct/sdcm/nemesis.py", line 446, in call_random_disrupt_method
  File "/sct/sdcm/nemesis.py", line 303, in disrupt_terminate_and_replace_node
  File "/sct/sdcm/nemesis.py", line 248, in _add_and_init_new_cluster_node
  File "/sct/sdcm/cluster.py", line 1386, in wrapper
AssertionError: Setup has failed!
```

--- a/sdcm/cluster.py
+++ b/sdcm/cluster.py
@@ -1484,6 +1484,14 @@ class BaseCluster(object):
         node.destroy()
 
 
+class NodeSetupFailed(Exception):
+    pass
+
+
+class NodeSetupTimeout(Exception):
+    pass
+
+
 def wait_for_init_wrap(method):
     """
     Wraps wait_for_init class method.
@@ -1525,7 +1533,8 @@ def wait_for_init_wrap(method):
             time_elapsed = time.time() - start_time
             try:
                 node, node_status = queue.get(block=True, timeout=5)
-                assert node_status is True, 'Setup has failed!'
+                if not node_status:
+                    raise NodeSetupFailed("{node}:{node_status}".format(**locals()))
                 results.append(node)
                 cl_inst.log.info("(%d/%d) nodes ready, node %s. Time elapsed: %d s",
                                  len(results), len(node_list), str(node), int(time_elapsed))
@@ -1535,7 +1544,7 @@ def wait_for_init_wrap(method):
                 msg = 'TIMEOUT [%d min]: Waiting for node(-s) setup(%d/%d) expired!' % (
                     timeout, len(results), len(node_list))
                 cl_inst.log.error(msg)
-                raise Exception(msg)
+                raise NodeSetupTimeout(msg)
 
         time_elapsed = time.time() - start_time
         cl_inst.log.debug('Setup duration -> %s s', int(time_elapsed))

--- a/sdcm/nemesis.py
+++ b/sdcm/nemesis.py
@@ -25,9 +25,9 @@ import threading
 
 from avocado.utils import process
 
-from sdcm.cluster import SCYLLA_YAML_PATH
+from sdcm.cluster import SCYLLA_YAML_PATH, NodeSetupTimeout, NodeSetupFailed
 from sdcm.mgmt import TaskStatus
-from .utils import get_data_dir_path
+from .utils import get_data_dir_path, retrying
 from .log import SDCMAdapter
 from .keystore import KeyStore
 from . import prometheus
@@ -259,12 +259,19 @@ class Nemesis(object):
             self.target_node.stop_scylla_server(verify_up=False, verify_down=True)
             self.target_node.start_scylla_server(verify_up=True, verify_down=False)
 
+    @retrying(n=3, sleep_time=60, allowed_exceptions=(NodeSetupFailed, NodeSetupTimeout))
     def _add_and_init_new_cluster_node(self, old_node_private_ip=None, timeout=10800):
         """When old_node_private_ip is not None replacement node procedure is initiated"""
         self.log.info("Adding new node to cluster...")
         new_node = self.cluster.add_nodes(count=1, dc_idx=self.target_node.dc_idx, enable_auto_bootstrap=True)[0]
         new_node.replacement_node_ip = old_node_private_ip
-        self.cluster.wait_for_init(node_list=[new_node], timeout=timeout)
+        try:
+            self.cluster.wait_for_init(node_list=[new_node], timeout=timeout)
+        except (NodeSetupFailed, NodeSetupTimeout):
+            self.log.warning("Setup of the '%s' failed, removing it from list of nodes" % new_node)
+            self.cluster.nodes.remove(new_node)
+            self.log.warning("Node will not be terminated. Please terminate manually!!!")
+            raise
         self.cluster.wait_for_nodes_up_and_normal(nodes=[new_node])
         self.monitoring_set.reconfigure_scylla_monitoring()
         return new_node

